src/expr/term_context.h

```diff
@@ -0,0 +1,143 @@
+/*********************                                                        */
+/*! \file term_context.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Term context utilities.
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__EXPR__TERM_CONTEXT_H
+#define CVC4__EXPR__TERM_CONTEXT_H
+
+#include "expr/node.h"
+
+namespace CVC4 {
+
+/**
+ * This is an abstract class for computing "term context identifiers". A term
+ * context identifier is a hash value that identifies some property of the
+ * context in which a term occurs. Common examples of the implementation of
+ * such a mapping are implemented in the subclasses below.
+ *
+ * A term context identifier is intended to be information that can be locally
+ * computed from the parent's hash, and hence does not rely on maintaining
+ * paths.
+ *
+ * In the below documentation, we write t @ [p] to a term at a given position,
+ * where p is a list of indices. For example, the atomic subterms of:
+ *   (and P (not Q))
+ * are P @ [0] and Q @ [1,0].
+ */
+class TermContext
+{
+ public:
+  TermContext() {}
+  virtual ~TermContext() {}
+  /** The default initial value of root terms. */
+  virtual uint32_t initialValue() const = 0;
+  /**
+   * Returns the term context identifier of the index^th child of t, where tval
+   * is the term context identifier of t.
+   */
+  virtual uint32_t computeValue(TNode t, uint32_t tval, size_t index) const = 0;
+};
+
+/**
+ * Remove term formulas (rtf) term context.
+ *
+ * Computes whether we are inside a term (as opposed to being part of Boolean
+ * skeleton) and whether we are inside a quantifier. For example, for:
+ *   (and (= a b) (forall ((x Int)) (P x)))
+ * we have the following mappings (term -> inTerm,inQuant)
+ *   (= a b) @ [0] -> false, false
+ *   a @ [0,1] -> true, false
+ *   (P x) @ [1,1] -> false, true
+ *    x @ [1,1,0] -> true, true
+ * Notice that the hash of a child can be computed from the parent's hash only,
+ * and hence this can be implemented without considering paths.
+ */
+class RtfTermContext : public TermContext
+{
+ public:
+  RtfTermContext() {}
+  /** The initial value: not in a term context or beneath a quantifier. */
+  uint32_t initialValue() const override;
+  /** Compute the value */
```

Comment:
Say abstractly what is the pattern of the values

Reply:
What do you mean?

Reply:
The value is computed in a given way, so that for a context like this you get a value like that. I think it'd be interesting to have an idea of that from this point.

Reply:
That's what the comment at the top of the class is for.  The user should look there instead.  

Notice that more generally any method marked `override` should not need significant documentation.

Reply:
Sorry, I don't mean how the polarity is determined but what is the structure of the hash.